Sample Models/Chemistry & Physics/Chaos in a Box.nlogo

```diff
@@ -0,0 +1,906 @@
+globals [
+  collision-precision ; Determines how accurate the location of collisions will be
+
+  ; Determines how much we round the angles when bouncing off the obstacle.
+  ; This is necessary, because small errors in angle should be ignored.
+  angle-precision
+
+  speed ; All balls have the same speed
+
+  ; These are used to track which agent is being dragged and what part of that agent
+  ; is being dragged.
+  target-agent
+  set-target-attribute
+  get-target-attribute
+  target-attribute-value
+]
+
+breed [ obstacles obstacle ]
+breed [ balls ball ]
+
+to setup [ obstacle-placement ball-placement ]
+  clear-all
+  set-default-shape obstacles "circle"
+
+  ; Create obstacle so that it's completely inside the world
+  create-obstacles num-obstacles [
+    set color grey
+    ; First, set the size of the obstacle so that it fits in the world and doesn't take up all the space
+    set size max list 1 random-float (0.75 * min list world-width world-height)
+    run obstacle-placement
+  ]
+
+  set-default-shape balls "circled-default"
+  create-balls 1 [
+    set color red
+    run ball-placement
+  ]
+
+  set collision-precision 64 ; Collisions will be within speed / (2 ^ 64) of the correct position
+  set angle-precision 1000 ; Round to 1000th of a degree
+  set speed 0.05
+
+  reset-dragging
+
+  reset-ticks
+end
+
+to setup-random
+  setup [-> place-randomly-inside-walls] [->
+    place-randomly-inside-walls
+    while [ any? obstacles with [ overlap myself > 0 ] ] [
+      place-randomly-inside-walls
+    ]
+  ]
+end
+
+
+to setup-periodic-x
+  set num-obstacles 1
+  setup [->
+    setxy 0 0
+    set size 3
+  ] [ ->
+    setxy 6 8
+    facexy 0 2
+  ]
+end
+
+to setup-periodic-quilt
+  set num-obstacles 1
+  setup [-> setxy 0 8 set size 1 ] [-> setxy 0 7 facexy -8 -6 ]
+end
+
+to add-ball
+  ask balls with [ color = red ] [
+    ; Place the second ball at almost, but not quite, the same position as the first
+    let new-color [ color ] of max-one-of balls [ who ] + 90
```

Comment:
We generally outlaw the use of `who` in models library models. I'm willing to make an exception since this case actually makes sense, but if you see an easy way of rewriting this one line, I would endorse it.